system/include/emscripten/val.h

```diff
@@ -648,6 +659,83 @@ inline val::iterator val::begin() const {
   return iterator(*this);
 }
 
+#if _LIBCPP_STD_VER >= 20
+namespace internal {
+struct val_awaiter {
+  val_awaiter(val&& promise)
+    : state(std::in_place_index<STATE_PROMISE>, std::move(promise)) {}
+
+  // just in case, ensure nobody moves / copies this type around
+  val_awaiter(val_awaiter&&) = delete;
+
+  bool await_ready() { return false; }
+
+  void await_suspend(std::coroutine_handle<val::promise_type> handle) {
+    internal::_emval_coro_suspend(std::get<STATE_PROMISE>(state).as_handle(), this);
+    state.emplace<STATE_CORO>(handle);
+  }
+
+  void resume_with(val&& result) {
+    auto coro = std::move(std::get<STATE_CORO>(state));
+    state.emplace<STATE_RESULT>(std::move(result));
+    coro.resume();
+  }
+
+  val await_resume() { return std::move(std::get<STATE_RESULT>(state)); }
+
+private:
+  // State machine holding awaiter's current state. One of:
+  //  - initially created with promise
+  //  - waiting with a given coroutine handle
+  //  - completed with a result
+  std::variant<val, std::coroutine_handle<val::promise_type>, val> state;
+
+  constexpr static std::size_t STATE_PROMISE = 0;
+  constexpr static std::size_t STATE_CORO = 1;
+  constexpr static std::size_t STATE_RESULT = 2;
```

Comment:
Can we move all this to the top of the class since understanding this state is necessary for understanding the rest of the class?

Reply:
Yeah TBH I don't understand why the common style is to put private items in the end of the class.